casper/src/main/scala/coop/rchain/casper/genesis/Genesis.scala

```diff
@@ -103,6 +94,62 @@ object Genesis {
                      )
     } yield genesisBlock
 
+  private def getVaults[F[_]: Sync: Log: RaiseIOError](
+      maybeVaultPath: Option[String],
+      defaultVaultPath: Path
+  ): F[Seq[Vault]] =
+    maybeVaultPath match {
+      case Some(vaultsPathStr) =>
+        val vaultsPath = Paths.get(vaultsPathStr)
+        Monad[F].ifM(exists(vaultsPath))(
+          vaultFromFile[F](vaultsPath),
+          Sync[F].raiseError(new Exception(s"Specified vaults file $vaultsPath does not exist"))
+        )
+      case None =>
+        Monad[F].ifM(exists(defaultVaultPath))(
+          Log[F].info(s"Using default file $defaultVaultPath") >> vaultFromFile[F](
+            defaultVaultPath
+          ),
+          Log[F]
+            .warn(
+              "No vaults file specified and no default file found. No vaults will exist at genesis."
+            )
+            .map(_ => Seq.empty[Vault])
+        )
+    }
+
+  private def vaultFromFile[F[_]: Sync: Log: RaiseIOError](vaultsPath: Path): F[Seq[Vault]] =
+    for {
+      lines <- SourceIO
+                .open(vaultsPath)
+                .use(_.getLines)
+                .adaptError {
+                  case ex: Throwable =>
+                    new RuntimeException(
+                      s"Failed to read ${vaultsPath.toAbsolutePath} for reason: ${ex.getMessage}"
+                    )
+                }
+      vaults <- lines.traverse(parseLine[F])
+    } yield vaults
+
+  private def parseLine[F[_]: Sync: Log: RaiseIOError](line: String): F[Vault] =
+    Sync[F].fromEither(
+      fromLine(line)
+        .leftMap(errMsg => new RuntimeException(s"Error in parsing vaults file: $errMsg"))
+    )
+
+  private def fromLine(line: String): Either[String, Vault] = line.split(",") match {
+    case Array(ethAddress, initRevBalanceStr, _) =>
```

Comment:
what's on the 3rd coordinate? Why do we ignore it?

Reply:
I don't think there is anything in the 3rd+ coordinate.

Reply:
`case Array(a, b, _)` means that you are matching three coordinates. In https://files.rchain-dev.tk/wallets_all.txt I can see that there are some 0s and 1s in the third coordinate. Do you know what do they stand for?

Reply:
Now I remember: 0 or 1 corresponds to whether that address is a contract. We need to handle those that are contracts in a special fashion - we probably need to have a larger discussion for this but any adjustments that result should go in a future PR.

Reply:
Please create a JIRA issue describing why they need special treatment, so that we don't forget to (desing) and implement it